bindgen/templates/Async.cs

```diff
@@ -0,0 +1,88 @@
+[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+delegate void UniFfiFutureCallback(IntPtr continuationHandle, byte pollResult);
+
+internal static class _UniFFIAsync {
+    internal const byte UNIFFI_RUST_FUTURE_POLL_READY = 0;
+    // internal const byte UNIFFI_RUST_FUTURE_POLL_MAYBE_READY = 1;
+
+    // FFI type for Rust future continuations
+    internal class UniffiRustFutureContinuationCallback
+    {
+        public static UniFfiFutureCallback callback = Callback;
+
+        public static void Callback(IntPtr continuationHandle, byte pollResult)
+        {
+            GCHandle handle = GCHandle.FromIntPtr(continuationHandle);
+            if (handle.Target is TaskCompletionSource<byte> tcs) 
+            {
+                tcs.SetResult(pollResult);
+            }
+            else 
+            {
+                throw new InternalException("Unable to cast unmanaged IntPtr to TaskCompletionSource<byte>");
+            }
+        }
+
+        public static void Register()
+        {
+            IntPtr fn = Marshal.GetFunctionPointerForDelegate(callback);
+            _UniFFILib.{{ ci.ffi_rust_future_continuation_callback_set().name() }}(fn);
+        }
+    }
+
+    public delegate F CompleteFuncDelegate<F>(IntPtr ptr, ref RustCallStatus status);
+
+    public delegate void CompleteActionDelegate(IntPtr ptr, ref RustCallStatus status);
+
+    private static async Task PollFuture(IntPtr rustFuture, Action<IntPtr, IntPtr> pollFunc)
+    {
+        byte pollResult;
+        do 
+        {
+            var tcs = new TaskCompletionSource<byte>();
```

Comment:
```suggestion
            var tcs = new TaskCompletionSource<byte>(TaskCreationOptions.RunContinuationsAsynchronously);
```

`TaskCompletionSource` constructed with default parameters can deadlock if `tcs.SetResult` is invoked from a non-managed thread. I came across this deadlock when I tried to run the tests on a single thread. By default, Xunit (the test framework used in this project) runs tests in parallel, I thought running the tests sequentially would  solve the flaky async tests. Running the tests sequentially didn't help with the flaky tests, but at least I discovered this deadlock 😅 

https://techyian.github.io/2019-07-16-taskcompletionsource-deadlock/

Reply:
Awesome! We were just investigating a deadlock issue, but we didn't know where it was coming from. This fixed it for us :)